File: whiteboardapp/src/wasm/whiteboard.cpp
#include "../../include/wasm/whiteboard.hpp"

// Line implementation
void Line::draw(emscripten::val context) {
    if (points.empty()) return;

    context.call<void>("beginPath");
    context.set("strokeStyle", color);
    context.set("lineWidth", thickness);
    context.set("lineCap", std::string("round"));
    context.set("lineJoin", std::string("round"));

    context.call<void>("moveTo", points[0].x, points[0].y);
    for (size_t i = 1; i < points.size(); i++) {
        context.call<void>("lineTo", points[i].x, points[i].y);
    }
    context.call<void>("stroke");

    // Draw selection outline if selected
    if (selected) {
        Rect bounds = getBounds();
        context.set("strokeStyle", "#0095ff");
        context.set("lineWidth", 2);
        context.call<void>("strokeRect", bounds.x - 5, bounds.y - 5, 
                          bounds.width + 10, bounds.height + 10);
    }
}

bool Line::containsPoint(float x, float y) {
    for (size_t i = 1; i < points.size(); i++) {
        float x1 = points[i-1].x, y1 = points[i-1].y;
        float x2 = points[i].x, y2 = points[i].y;
        
        float distance = std::abs((y2-y1)*x - (x2-x1)*y + x2*y1 - y2*x1) /
                        std::sqrt((y2-y1)*(y2-y1) + (x2-x1)*(x2-x1));
        
        if (distance < 5.0f) return true;
    }
    return false;
}

void Line::move(float dx, float dy) {
    for (auto& point : points) {
        point.x += dx;
        point.y += dy;
    }
}

Rect Line::getBounds() {
    if (points.empty()) return {0, 0, 0, 0};
    
    float minX = points[0].x, maxX = points[0].x;
    float minY = points[0].y, maxY = points[0].y;
    
    for (const auto& point : points) {
        minX = std::min(minX, point.x);
        maxX = std::max(maxX, point.x);
        minY = std::min(minY, point.y);
        maxY = std::max(maxY, point.y);
    }
    
    return {minX, minY, maxX - minX, maxY - minY};
}

// Rectangle implementation
void Rectangle::draw(emscripten::val context) {
    context.set("strokeStyle", color);
    context.set("lineWidth", thickness);
    context.call<void>("strokeRect", bounds.x, bounds.y, bounds.width, bounds.height);

    if (selected) {
        context.set("strokeStyle", "#0095ff");
        context.set("lineWidth", 2);
        context.call<void>("strokeRect", bounds.x - 5, bounds.y - 5, 
                          bounds.width + 10, bounds.height + 10);
    }
}

bool Rectangle::containsPoint(float x, float y) {
    return x >= bounds.x && x <= bounds.x + bounds.width &&
           y >= bounds.y && y <= bounds.y + bounds.height;
}

void Rectangle::move(float dx, float dy) {
    bounds.x += dx;
    bounds.y += dy;
}

// Circle implementation
void Circle::draw(emscripten::val context) {
    context.call<void>("beginPath");
    context.set("strokeStyle", color);
    context.set("lineWidth", thickness);
    context.call<void>("arc", center.x, center.y, radius, 0, 2 * M_PI);
    context.call<void>("stroke");

    if (selected) {
        context.set("strokeStyle", "#0095ff");
        context.set("lineWidth", 2);
        Rect bounds = getBounds();
        context.call<void>("strokeRect", bounds.x - 5, bounds.y - 5, 
                          bounds.width + 10, bounds.height + 10);
    }
}

bool Circle::containsPoint(float x, float y) {
    float dx = x - center.x;
    float dy = y - center.y;
    return (dx * dx + dy * dy) <= (radius * radius);
}

void Circle::move(float dx, float dy) {
    center.x += dx;
    center.y += dy;
}

Rect Circle::getBounds() {
    return {center.x - radius, center.y - radius, radius * 2, radius * 2};
}

// Whiteboard implementation
Whiteboard::Whiteboard() 
    : currentColor("#000000"), currentThickness(2.0f), 
      currentShape(ShapeType::FREEHAND), isSelecting(false) {
    init();
}

void Whiteboard::init() {
    elements.clear();
    selectedElements.clear();
    currentElement = nullptr;
    isSelecting = false;
}

void Whiteboard::startDrawing(float x, float y) {
    if (isSelecting) {
        startSelection(x, y);
        return;
    }

    switch (currentShape) {
        case ShapeType::FREEHAND: {
            auto line = std::make_shared<Line>();
            line->points.push_back({x, y});
            line->color = currentColor;
            line->thickness = currentThickness;
            currentElement = line;
            break;
        }
        case ShapeType::RECTANGLE: {
            auto rect = std::make_shared<Rectangle>();
            rect->bounds = {x, y, 0, 0};
            rect->color = currentColor;
            rect->thickness = currentThickness;
            currentElement = rect;
            break;
        }
        case ShapeType::CIRCLE: {
            auto circle = std::make_shared<Circle>();
            circle->center = {x, y};
            circle->radius = 0;
            circle->color = currentColor;
            circle->thickness = currentThickness;
            currentElement = circle;
            break;
        }
        default:
            break;
    }

    if (currentElement) {
        elements.push_back(currentElement);
    }
}

void Whiteboard::continueDrawing(float x, float y) {
    if (isSelecting) {
        updateSelection(x, y);
        return;
    }

    if (!currentElement) return;

    switch (currentShape) {
        case ShapeType::FREEHAND: {
            auto line = std::static_pointer_cast<Line>(currentElement);
            line->points.push_back({x, y});
            break;
        }
        case ShapeType::RECTANGLE: {
            auto rect = std::static_pointer_cast<Rectangle>(currentElement);
            float dx = x - rect->bounds.x;
            float dy = y - rect->bounds.y;
            rect->bounds.width = dx;
            rect->bounds.height = dy;
            break;
        }
        case ShapeType::CIRCLE: {
            auto circle = std::static_pointer_cast<Circle>(currentElement);
            float dx = x - circle->center.x;
            float dy = y - circle->center.y;
            circle->radius = std::sqrt(dx * dx + dy * dy);
            break;
        }
        default:
            break;
    }
}

void Whiteboard::endDrawing() {
    if (isSelecting) {
        endSelection();
        return;
    }

    currentElement = nullptr;
}

void Whiteboard::draw(emscripten::val context) {
    for (const auto& element : elements) {
        element->draw(context);
    }
}

void Whiteboard::setShapeType(ShapeType type) {
    currentShape = type;
    clearSelection();
}

void Whiteboard::setColor(const std::string& color) {
    currentColor = color;
}

void Whiteboard::setThickness(float thickness) {
    currentThickness = thickness;
}

void Whiteboard::startSelection(float x, float y) {
    selectionStart = {x, y};
    clearSelection();
}

void Whiteboard::updateSelection(float x, float y) {
    float left = std::min(x, selectionStart.x);
    float top = std::min(y, selectionStart.y);
    float right = std::max(x, selectionStart.x);
    float bottom = std::max(y, selectionStart.y);

    for (auto& element : elements) {
        Rect bounds = element->getBounds();
        bool intersects = !(bounds.x > right || 
                          bounds.x + bounds.width < left ||
                          bounds.y > bottom || 
                          bounds.y + bounds.height < top);
        element->selected = intersects;
        if (intersects) {
            selectedElements.push_back(element);
        }
    }
}

void Whiteboard::endSelection() {
    isSelecting = false;
}

void Whiteboard::moveSelected(float dx, float dy) {
    for (auto& element : selectedElements) {
        element->move(dx, dy);
    }
}

void Whiteboard::deleteSelected() {
    elements.erase(
        std::remove_if(elements.begin(), elements.end(),
            [](const auto& element) { return element->selected; }),
        elements.end()
    );
    selectedElements.clear();
}

void Whiteboard::clearSelection() {
    for (auto& element : elements) {
        element->selected = false;
    }
    selectedElements.clear();
}

void Whiteboard::clear() {
    init();
}

void Whiteboard::erase(float x, float y, float radius) {
    elements.erase(
        std::remove_if(elements.begin(), elements.end(),
            [x, y, radius](const auto& element) {
                return element->containsPoint(x, y);
            }),
        elements.end()
    );
}

// Binding code for Emscripten
EMSCRIPTEN_BINDINGS(whiteboard_module) {
    emscripten::enum_<ShapeType>("ShapeType")
        .value("FREEHAND", ShapeType::FREEHAND)
        .value("RECTANGLE", ShapeType::RECTANGLE)
        .value("CIRCLE", ShapeType::CIRCLE)
        .value("LINE", ShapeType::LINE)
        .value("TRIANGLE", ShapeType::TRIANGLE);

    emscripten::class_<Whiteboard>("Whiteboard")
        .constructor<>()
        .function("init", &Whiteboard::init)
        .function("startDrawing", &Whiteboard::startDrawing)
        .function("continueDrawing", &Whiteboard::continueDrawing)
        .function("endDrawing", &Whiteboard::endDrawing)
        .function("setShapeType", &Whiteboard::setShapeType)
        .function("setColor", &Whiteboard::setColor)
        .function("setThickness", &Whiteboard::setThickness)
        .function("draw", &Whiteboard::draw)
        .function("clear", &Whiteboard::clear)
        .function("erase", &Whiteboard::erase)
        .function("startSelection", &Whiteboard::startSelection)
        .function("updateSelection", &Whiteboard::updateSelection)
        .function("endSelection", &Whiteboard::endSelection)
        .function("moveSelected", &Whiteboard::moveSelected)
        .function("deleteSelected", &Whiteboard::deleteSelected)
        .function("clearSelection", &Whiteboard::clearSelection);
} 

File: whiteboardapp/src/lib/whiteboard.ts
/**
 * @file whiteboard.ts
 * @brief TypeScript wrapper for the WebAssembly whiteboard implementation
 * 
 * This file provides the bridge between the C++ WebAssembly code and the React frontend.
 * It handles:
 * - WebAssembly module loading and initialization
 * - Canvas event management
 * - Drawing state management
 * - Touch and mouse input handling
 */

/**
 * @brief Shape types available for drawing
 * 
 * These values must match the C++ ShapeType enum exactly.
 * Used to communicate the current drawing tool between TypeScript and C++.
 */
export enum ShapeType {
    FREEHAND = 'FREEHAND',   // Freehand drawing mode
    RECTANGLE = 'RECTANGLE', // Rectangle shape tool
    CIRCLE = 'CIRCLE',      // Circle shape tool
    LINE = 'LINE',          // Straight line tool
    TRIANGLE = 'TRIANGLE'   // Triangle shape tool (future use)
}

/**
 * @brief Interface for the WebAssembly module's Whiteboard constructor
 * 
 * This interface defines the shape of the module exported by Emscripten.
 * It provides access to the Whiteboard class constructor and shape type enum.
 */
interface WhiteboardModule {
    Whiteboard: {
        new(): Whiteboard;
    };
    ShapeType: typeof ShapeType;
}

/**
 * @brief Interface for the WebAssembly-compiled Whiteboard class
 * 
 * This interface matches the public methods of the C++ Whiteboard class.
 * It defines all operations that can be performed on the whiteboard.
 */
interface Whiteboard {
    init(): void;                                    // Initialize/reset the whiteboard
    startDrawing(x: number, y: number): void;        // Start a drawing operation
    continueDrawing(x: number, y: number): void;     // Continue current drawing
    endDrawing(): void;                             // End current drawing
    setShapeType(type: ShapeType): void;            // Set the current shape tool
    setColor(color: string): void;                  // Set drawing color
    setThickness(thickness: number): void;          // Set line thickness
    draw(context: CanvasRenderingContext2D): void;  // Draw to canvas
    clear(): void;                                  // Clear the canvas
    erase(x: number, y: number, radius: number): void; // Erase at point
    startSelection(x: number, y: number): void;     // Start selection operation
    updateSelection(x: number, y: number): void;    // Update selection area
    endSelection(): void;                          // End selection operation
    moveSelected(dx: number, dy: number): void;     // Move selected elements
    deleteSelected(): void;                        // Delete selected elements
    clearSelection(): void;                        // Clear selection state
    redraw(): void;
}

// Module cache to prevent reloading the WebAssembly module
let wasmModule: any = null;

/**
 * @brief Tool types available for drawing
 * 
 * These values must match the C++ Tool enum exactly.
 * Used to communicate the current drawing tool between TypeScript and C++.
 */
export enum Tool {
    DRAW = 'DRAW',
    SELECT = 'SELECT',
    ERASE = 'ERASE'
}

/**
 * @brief Main wrapper class for the WebAssembly whiteboard
 * 
 * This class provides a TypeScript-friendly interface to the WebAssembly module.
 * It handles:
 * - Module initialization
 * - Event management
 * - Canvas interaction
 * - Drawing state
 */
export class WhiteboardWrapper {
    private module: WhiteboardModule | null = null;        // WebAssembly module reference
    private whiteboard: Whiteboard | null = null;          // Whiteboard instance
    private canvas: HTMLCanvasElement | null = null;       // Canvas element
    private context: CanvasRenderingContext2D | null = null; // Canvas context
    private isDrawing = false;                            // Drawing state flag
    private lastX = 0;                                    // Last mouse/touch X position
    private lastY = 0;                                    // Last mouse/touch Y position
    private currentShape: ShapeType = ShapeType.FREEHAND; // Current shape tool
    private currentTool: Tool = Tool.DRAW;                // Current tool
    private eraserRadius = 20; // Default eraser size
    private dragStartX = 0;
    private dragStartY = 0;
    private isDraggingSelection = false;

    /**
     * @brief Initialize the whiteboard with a canvas element
     * 
     * This method:
     * 1. Sets up the canvas and context
     * 2. Loads the WebAssembly module
     * 3. Initializes the whiteboard
     * 4. Sets up event listeners
     * 
     * @param canvas HTML Canvas element to draw on
     * @throws Error if canvas context cannot be obtained
     * @throws Error if WebAssembly module fails to load
     */
    async initialize(canvas: HTMLCanvasElement) {
        this.canvas = canvas;
        this.context = canvas.getContext('2d');
        
        if (!this.context) {
            throw new Error('Could not get canvas context');
        }

        try {
            // Load WebAssembly module if not already loaded
            if (!wasmModule) {
                const { default: createModule } = await import(/* webpackIgnore: true */ '/wasm/whiteboard.js');
                wasmModule = await createModule({
                    locateFile: (path: string) => {
                        // Handle WebAssembly file location
                        if (path.endsWith('.wasm')) {
                            return '/wasm/whiteboard.wasm';
                        }
                        return path;
                    },
                });
            }

            // Initialize whiteboard instance
            this.module = wasmModule;
            this.whiteboard = new this.module.Whiteboard();
            this.whiteboard.init();
            this.setupEventListeners();
        } catch (error) {
            console.error('Failed to initialize WebAssembly module:', error);
            throw error;
        }
    }

    /**
     * @brief Set up canvas event listeners
     * 
     * Adds listeners for:
     * - Mouse events (desktop)
     * - Touch events (mobile)
     * - Pointer events (future use)
     */
    private setupEventListeners() {
        if (!this.canvas) return;

        // Mouse event listeners
        this.canvas.addEventListener('mousedown', this.handleMouseDown);
        this.canvas.addEventListener('mousemove', this.handleMouseMove);
        this.canvas.addEventListener('mouseup', this.handleMouseUp);
        this.canvas.addEventListener('mouseleave', this.handleMouseUp);

        // Touch event listeners
        this.canvas.addEventListener('touchstart', this.handleTouchStart);
        this.canvas.addEventListener('touchmove', this.handleTouchMove);
        this.canvas.addEventListener('touchend', this.handleTouchEnd);
    }

    /**
     * @brief Handle mouse down event
     * 
     * Starts drawing or selection operation based on current mode.
     * Converts window coordinates to canvas coordinates.
     */
    private handleMouseDown = (e: MouseEvent) => {
        if (!this.whiteboard || !this.canvas) return;

        const rect = this.canvas.getBoundingClientRect();
        const x = e.clientX - rect.left;
        const y = e.clientY - rect.top;

        this.isDrawing = true;
        this.lastX = x;
        this.lastY = y;
        this.dragStartX = x;
        this.dragStartY = y;

        switch (this.currentTool) {
            case Tool.DRAW:
                this.whiteboard.startDrawing(x, y);
                break;
            case Tool.SELECT:
                if (e.shiftKey) {
                    // Start selection box
                    this.whiteboard.startSelection(x, y);
                    this.isDraggingSelection = false;
                } else {
                    // Start dragging selected shapes
                    this.isDraggingSelection = true;
                }
                break;
            case Tool.ERASE:
                this.whiteboard.erase(x, y, this.eraserRadius);
                break;
        }
        this.draw();
    };

    /**
     * @brief Handle mouse move event
     * 
     * Continues drawing or updates selection based on current mode.
     * Converts window coordinates to canvas coordinates.
     */
    private handleMouseMove = (e: MouseEvent) => {
        if (!this.whiteboard || !this.canvas) return;

        const rect = this.canvas.getBoundingClientRect();
        const x = e.clientX - rect.left;
        const y = e.clientY - rect.top;

        // Always draw eraser circle if in erase mode
        if (this.currentTool === Tool.ERASE) {
            this.drawEraserCircle(x, y);
        }

        if (!this.isDrawing) return;

        switch (this.currentTool) {
            case Tool.DRAW:
                this.whiteboard.continueDrawing(x, y);
                break;
            case Tool.SELECT:
                if (!this.isDraggingSelection) {
                    // Update selection box
                    this.whiteboard.updateSelection(x, y);
                } else {
                    // Move selected shapes
                    const dx = x - this.lastX;
                    const dy = y - this.lastY;
                    this.whiteboard.moveSelected(dx, dy);
                }
                break;
            case Tool.ERASE:
                this.whiteboard.erase(x, y, this.eraserRadius);
                break;
        }

        this.lastX = x;
        this.lastY = y;
        this.draw();
    };

    /**
     * @brief Handle mouse up event
     * 
     * Ends drawing or selection operation based on current mode.
     */
    private handleMouseUp = () => {
        if (!this.whiteboard || !this.isDrawing) return;

        this.isDrawing = false;
        this.isDraggingSelection = false;

        switch (this.currentTool) {
            case Tool.DRAW:
                this.whiteboard.endDrawing();
                break;
            case Tool.SELECT:
                this.whiteboard.endSelection();
                break;
        }
        this.draw();
    };

    /**
     * @brief Handle touch start event
     * 
     * Starts drawing or selection operation based on current mode.
     * Converts window coordinates to canvas coordinates.
     */
    private handleTouchStart = (e: TouchEvent) => {
        if (!this.whiteboard || !this.canvas) return;
        e.preventDefault();

        const rect = this.canvas.getBoundingClientRect();
        const touch = e.touches[0];
        const x = touch.clientX - rect.left;
        const y = touch.clientY - rect.top;

        this.isDrawing = true;
        this.lastX = x;
        this.lastY = y;
        this.dragStartX = x;
        this.dragStartY = y;

        switch (this.currentTool) {
            case Tool.DRAW:
                this.whiteboard.startDrawing(x, y);
                break;
            case Tool.SELECT:
                if (e.touches.length === 2) {
                    // Two-finger touch starts selection
                    this.whiteboard.startSelection(x, y);
                    this.isDraggingSelection = false;
                } else {
                    // Single-finger touch starts dragging
                    this.isDraggingSelection = true;
                }
                break;
            case Tool.ERASE:
                this.whiteboard.erase(x, y, this.eraserRadius);
                break;
        }
        this.draw();
    };

    /**
     * @brief Handle touch move event
     * 
     * Continues drawing or updates selection based on current mode.
     * Converts window coordinates to canvas coordinates.
     */
    private handleTouchMove = (e: TouchEvent) => {
        if (!this.whiteboard || !this.canvas || !this.isDrawing) return;
        e.preventDefault();

        const rect = this.canvas.getBoundingClientRect();
        const touch = e.touches[0];
        const x = touch.clientX - rect.left;
        const y = touch.clientY - rect.top;

        switch (this.currentTool) {
            case Tool.DRAW:
                this.whiteboard.continueDrawing(x, y);
                break;
            case Tool.SELECT:
                if (!this.isDraggingSelection) {
                    this.whiteboard.updateSelection(x, y);
                } else {
                    const dx = x - this.lastX;
                    const dy = y - this.lastY;
                    this.whiteboard.moveSelected(dx, dy);
                }
                break;
            case Tool.ERASE:
                this.whiteboard.erase(x, y, this.eraserRadius);
                break;
        }

        this.lastX = x;
        this.lastY = y;
        this.draw();
    };

    /**
     * @brief Handle touch end event
     * 
     * Ends drawing or selection operation based on current mode.
     */
    private handleTouchEnd = (e: TouchEvent) => {
        e.preventDefault();
        if (!this.whiteboard || !this.isDrawing) return;

        this.isDrawing = false;
        this.isDraggingSelection = false;

        switch (this.currentTool) {
            case Tool.DRAW:
                this.whiteboard.endDrawing();
                break;
            case Tool.SELECT:
                this.whiteboard.endSelection();
                break;
        }
        this.draw();
    };

    /**
     * @brief Set the current shape type
     * @param shape Shape type to use for drawing
     */
    setShape(shape: ShapeType) {
        if (!this.whiteboard) return;
        this.currentShape = shape;
        this.whiteboard.setShapeType(shape);
    }

    /**
     * @brief Set the current tool
     * @param tool Tool type to use for drawing
     */
    setTool(tool: Tool) {
        this.currentTool = tool;
        if (this.canvas) {
            switch (tool) {
                case Tool.SELECT:
                    this.canvas.style.cursor = 'crosshair';
                    break;
                case Tool.ERASE:
                    this.canvas.style.cursor = 'none';
                    break;
                case Tool.DRAW:
                    this.canvas.style.cursor = this.currentShape === ShapeType.FREEHAND ? 'default' : 'crosshair';
                    break;
            }
        }
    }

    /**
     * @brief Set the eraser size
     * @param size Eraser size in pixels
     */
    setEraserSize(size: number) {
        this.eraserRadius = size;
    }

    /**
     * @brief Move selected elements
     * @param dx X offset in pixels
     * @param dy Y offset in pixels
     */
    moveSelection(dx: number, dy: number) {
        if (!this.whiteboard) return;
        this.whiteboard.moveSelected(dx, dy);
        this.draw();
    }

    /**
     * @brief Delete selected elements
     */
    deleteSelection() {
        if (!this.whiteboard) return;
        this.whiteboard.deleteSelected();
        this.draw();
    }

    /**
     * @brief Set the current drawing color
     * @param color Color in HTML format (e.g., "#000000")
     */
    setColor(color: string) {
        if (!this.whiteboard) return;
        this.whiteboard.setColor(color);
    }

    /**
     * @brief Set the current line thickness
     * @param thickness Thickness in pixels
     */
    setThickness(thickness: number) {
        if (!this.whiteboard) return;
        this.whiteboard.setThickness(thickness);
    }

    /**
     * @brief Clear the entire canvas
     */
    clear() {
        if (!this.whiteboard || !this.context || !this.canvas) return;
        this.whiteboard.clear();
        this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
    }

    /**
     * @brief Erase content at specified coordinates
     * @param x X coordinate of eraser center
     * @param y Y coordinate of eraser center
     * @param radius Eraser radius in pixels
     */
    erase(x: number, y: number, radius: number) {
        if (!this.whiteboard) return;
        this.whiteboard.erase(x, y, radius);
        this.draw();
    }

    /**
     * @brief Draw the current state to the canvas
     * 
     * Clears the canvas and redraws all elements.
     * Called after each operation that modifies the drawing.
     */
    private draw() {
        if (!this.whiteboard || !this.context || !this.canvas) return;
        this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
        this.whiteboard.draw(this.context);
    }

    /**
     * @brief Draw the eraser circle on the canvas
     * @param x X coordinate of eraser center
     * @param y Y coordinate of eraser center
     */
    private drawEraserCircle(x: number, y: number) {
        if (!this.context || !this.canvas) return;

        // Redraw the main content
        this.draw();

        // Draw eraser circle
        this.context.beginPath();
        this.context.arc(x, y, this.eraserRadius, 0, Math.PI * 2);
        this.context.strokeStyle = '#000000';
        this.context.lineWidth = 1;
        this.context.stroke();
    }

    /**
     * @brief Create a shape at the specified position
     * @param shape Shape type to create
     * @param x X coordinate for shape center
     * @param y Y coordinate for shape center
     */
    createShape(shape: ShapeType, x: number, y: number) {
        if (!this.whiteboard) return;
        
        // Set the current shape type
        this.currentShape = shape;
        this.whiteboard.setShapeType(shape);
        
        // Create the shape
        this.whiteboard.startDrawing(x, y);
        this.whiteboard.endDrawing();
        
        // Redraw to show the new shape
        this.draw();
    }

    /**
     * @brief Handle canvas drag over event
     * @param e Drag event
     */
    private handleCanvasDragOver = (e: DragEvent) => {
        e.preventDefault();
        if (this.canvas) {
            this.canvas.style.cursor = 'copy';
        }
    };

    /**
     * @brief Handle canvas drop event
     * @param e Drop event
     */
    private handleCanvasDrop = (e: DragEvent) => {
        e.preventDefault();
        if (!this.whiteboard || !this.canvas) return;

        const rect = this.canvas.getBoundingClientRect();
        const x = e.clientX - rect.left;
        const y = e.clientY - rect.top;
        const shape = e.dataTransfer?.getData('shape') as ShapeType;

        if (shape) {
            this.createShape(shape, x, y);
        }

        if (this.canvas) {
            this.canvas.style.cursor = 'default';
        }
    };

    /**
     * @brief Get SVG representation of the current drawing
     * @returns SVG content as a string
     */
    getSVGContent(): string {
        if (!this.whiteboard || !this.canvas) return '';
        
        let svgContent = '';
        
        // Get SVG paths from the whiteboard
        const paths = this.whiteboard.getSVGPaths();
        if (paths) {
            svgContent = paths;
        }
        
        return svgContent;
    }

    redraw(): void {
        if (this.whiteboard && this.context) {
            // Clear the canvas
            this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
            // Redraw all shapes
            this.whiteboard.redraw();
        }
    }
} 